nft-client/src/cts1.rs

```diff
@@ -0,0 +1,377 @@
+            byte.serial(out)?;
+        }
+        Ok(())
+    }
+}
+
+/// Parse the additional data from a hex string.
+impl FromStr for ReceiveHookData {
+    type Err = hex::FromHexError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(ReceiveHookData(hex::decode(s)?.to_vec()))
+    }
+}
+
+/// A description of a transfer as according to the CTS1 specification.
+#[derive(Debug)]
+pub struct Transfer {
+    pub token_id:     TokenIdVec,
+    pub amount:       u64,
+    pub from:         Address,
+    pub to:           Address,
+    pub receive_name: Option<OwnedReceiveName>,
+    pub data:         ReceiveHookData,
+}
+
+/// Serialization of a transfer, as according to the CTS1 specification.
+impl Serial for Transfer {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        self.token_id.serial(out)?;
+        self.amount.serial(out)?;
+        self.from.serial(out)?;
+        self.to.serial(out)?;
+        if let Address::Contract(_) = self.to {
+            self.receive_name
+                .as_ref()
+                .ok_or_else(W::Err::default)?
+                .serial(out)?;
+        }
+        self.data.serial(out)?;
+        Ok(())
+    }
+}
+
+/// The parameter type for the NFT contract function `CTS1-NFT.transfer`.
+#[derive(Debug)]
+pub struct TransferParams(pub Vec<Transfer>);
+
+/// Serialization of the transfer parameter, as according to the CTS1
+/// specification.
+impl Serial for TransferParams {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        let len = u8::try_from(self.0.len()).map_err(|_| W::Err::default())?;
+        len.serial(out)?;
+        for transfer in &self.0 {
+            transfer.serial(out)?;
+        }
+        Ok(())
+    }
+}
+
+/// The type of update for an operator update.
+pub enum OperatorUpdate {
+    /// Remove the operator.
+    Remove,
+    /// Add an address as an operator.
+    Add,
+}
+
+/// The deserialization of an operator update as according to the CTS1
+/// specification.
+impl Deserial for OperatorUpdate {
+    fn deserial<R: Read>(source: &mut R) -> Result<Self, ParseError> {
+        let discriminant = source.read_u8()?;
+        match discriminant {
+            0 => Ok(OperatorUpdate::Remove),
+            1 => Ok(OperatorUpdate::Add),
+            _ => Err(ParseError::default()),
+        }
+    }
+}
+
+/// A SHA256 digest.
+pub type Sha256 = [u8; 32];
+
+/// A URL for the metadata.
+pub struct MetadataUrl {
+    /// The url encoded as according to CTS1.
+    pub url:  String,
+    /// An optional checksum of the content found at the URL.
+    pub hash: Option<Sha256>,
+}
+
+/// Deserialization for MetadataUrl as according to the CTS1 specification.
+impl Deserial for MetadataUrl {
+    fn deserial<R: Read>(source: &mut R) -> Result<Self, ParseError> {
+        let len = source.read_u16()?;
+        let mut bytes = Vec::new();
+        for _ in 0..len {
+            bytes.push(source.read_u8()?)
+        }
```

Comment:
This is very inefficient, why not use `deserial_string` from crypto_common?

Reply:
The Read trait in this implementation is from `concordium_contracts_common` and cannot be used with `deserial_string` from `crypto_common`.